### PR TITLE
Introduces a CGAL_destructor_assertion macro.

### PR DESCRIPTION
**Pull request by GilesBathgate**
**The problem:**  Throwing an exception from a destructor can cause problems since the callee is unable to catch an exception raised while the stack is unwinding. The result is the application will terminate with a hard crash.

**Consequence:** Since the exception is uncatchable, the program will terminate, but also the real exception is hidden by the the destructor exception.

**Solution:** Its desirable to not throw exceptions from the destructor if and only if an existing exception has not been caught, and the stack is unwinding. For this purpose I propose  a new macro that can safely be called from a destructor, even when the stack is currently unwinding. The macro makes use of `std::uncaught_exception()` and results in a nop when this is the case.

Additional discussions about this issue and what problems it can cause for client applications (rapcad and openscad) is detailed here: https://github.com/openscad/openscad/issues/410, there was also a discussion about the issue here: http://cgal-discuss.949826.n4.nabble.com/Uncatchable-exception-converting-from-Nef-polyhedron-to-Polyhedron-3-td4658603.html

Another instance of the same problem present in the openscad project: https://github.com/openscad/openscad/issues/1455

### CODE REVIEW

**Comment by lrineau** [comment]
@GilesBathgate Could you please rebase your branch from `cgal/master`?

@sloriot I find the patch good and useful. Could you also have a look?

**TODO:** I think that however the pull-request should also modify the documentation. The file to modify is [`Documentation/doc/Documentation/Developer_manual/Chapter_checks.txt`][1]

[1]: https://github.com/CGAL/cgal/blob/master/Documentation/doc/Documentation/Developer_manual/Chapter_checks.txt

**Comment by gdamiand** [comment]
Instead of disabling totally the macro, could it be better to display a message about the problem when std::uncaught_exception() ?

**Comment by GilesBathgate** [comment]
@gdamiand Although this does disable the macro from throwing an exception when the stack is unwinding, the uncaught exception still propagates, and so you still get an exception message which can be handled gracefully. If you are saying that the uncaught exception should be caught within the macro, some additional information added to it, and then rethrown. I guess that might work.

**Comment by gdamiand** [comment]
Sorry, but I don't understand how this works. 

If I am in a destructor and I use the new macro CGAL_destructor_assertion(EX), then std::uncaught_exception() is true ?

If yes, then the CGAL_destructor_assertion macro is disabled and no assertion are raised.

If no, assertion is raised like previously.

**Comment by GilesBathgate** [comment]
@gdamiand There are two scenarios.

Scenario A: No exceptions were thrown by the class, but the assertion in the destructor is false. In this case it is fine to throw an exception from the destructor.

Scenario B: The class threw an exception in a method. As part of the exception clean-up routines the program unwinds the stack, and deletes local variables as it does so. When it tries to delete the class that also fails the assertion, it is now in a situation where the current exception has not been caught, (and there is no chance of catching it the destructor either because it exists on different stack frame). At this point if you throw another exception the program terminates, (via std::terminate())

I admit its a little tricky to grasp,I found this article quite useful: http://www.kolpackov.net/projects/c++/eh/dtor-1.xhtml

The example given here http://en.cppreference.com/w/cpp/error/uncaught_exception also illustrates the scenarios I described above.

**Comment by GilesBathgate** [comment]
@gdamiand Here is a demonstration program:

```C++
#include <iostream>
#include <exception>
#include <stdexcept>

class Foo {
int check_protocoll;
public:
    Foo() : check_protocoll(0){}
    ~Foo() {
        // if(!std::uncaught_exception())
              throw std::runtime_error("Destructor assertion failed");
    }

    void success()
    {
        check_protocoll=1;
    }

    void fail() {
       throw std::runtime_error("Assertion failed");
    }
};

int main()
{
    try {
        Foo f;
        f.success();
    } catch (const std::exception& e) {
        std::cout << "Exception caught: " << e.what() << '\n';
    }
    try {
        Foo f;
        f.fail();
    } catch (const std::exception& e) {
        std::cout << "Exception caught: " << e.what() << '\n';
    }
    std::cout << "Exited program cleanly\n";
}
```

**Comment by gdamiand** [comment]
Ok, I understand now. Thank you very much for the detailed answer.

**Comment by GilesBathgate** [comment]
It may also be worth noting that when the above program is compiled for c++11, even Scenario A terminates the program. This is because in c++11 you have to explicitly mark destructor's with noexcept(false), if they are expected to throw exceptions.

**Comment by GilesBathgate** [comment]
@lrineau I propose an additional macro CGAL_NOEXCEPT(false), which will only be defined when CGAL_CXX11 is defined. This is to take care of the situation in my previous comment. Should I put that in another merge request or add it to this one?

**Comment by lrineau** [comment]
@GilesBathgate You can add it to this pull-request. It should also be documented in the developers manual.

**Comment by GilesBathgate** [comment]
@lrineau I am trying to put together some notes for the documentation. This is a first draft, but I am not an excellent technical documenter.

...
+ **Destructor assertions** These are checks which can be made within normal clean up of an object. A special macro is provided to ensure these checks are not made when the object is being cleaned up during exception handling.

The according macro names all have the format CGAL_&lt;check_type&gt; where &lt;check_type&gt; can be one of

+ precondition
+ postcondition
+ assertion
+ warning
+ static_assertion
+ destructor_assertion
...
## Exception handling

Any destructor which might throw an exception, including a destructor which uses the destructor_assertion macro, should be marked with the CGAL_NOEXCEPT(false) macro. This macro provides future compatibility with C++11 and above which provides the noexcept keyword.

**Comment by lrineau** [comment]
@GilesBathgate, that looks good. I suggest a small modification: explicit the name of the special macro.
> A special macro `CGAL_destructor_assertion`is provided to ensure...

**Comment by GilesBathgate** [comment]
@lrineau I have updated the documentation in this merge request now. Is there anything else you would like doing before lifting the TODO flag?

**Comment by lrineau** [comment]
@GilesBathgate Nothing, as far as I can see. It looks good.

@sloriot will now merge your pull-request in our testsuite branch, to test if it can be integrated in `cgal/master`.

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-65

File: Installation/include/CGAL/config.h [hunk 1]
```diff
@@ -463,4 +463,11 @@ typedef const void * Nullptr_t;   // Anticipate C++0x's std::nullptr_t
 
 } //namespace CGAL
 
+//Support for c++11 noexcept
+#ifdef CGAL_CXX11
```
Comment (sloriot) [inline Installation/include/CGAL/config.h]: This line should be replaced by:
`#if (BOOST_VERSION > 105000 && !defined(BOOST_NO_CXX11_NOEXCEPT) ) ||
      (BOOST_VERSION > 104600 && BOOST_VERSION < 105100 && !defined(BOOST_NO_NOEXCEPT) )`

Comment (lrineau) [inline Installation/include/CGAL/config.h]: It think that can be simplified to:
```C++
#if BOOST_VERSION > 104600 && !defined(BOOST_NO_CXX11_NOEXCEPT) && !defined(BOOST_NO_NOEXCEPT)
```

Comment (sloriot) [inline Installation/include/CGAL/config.h]: indeed

**Comment by GilesBathgate** [comment]
I am not sure I understand the intention of the  !defined (BOOST_NOEXCEPT) part. 

BOOST_NO_CXX11_NOEXCEPT is the define that determines whether CXX11 noexcept is available, whereas BOOST_NOEXCEPT, is boosts equivalent of the CGAL_NOEXCEPT macro which was introduced. See: http://www.boost.org/doc/libs/1_54_0/boost/chrono/config.hpp

**Comment by lrineau** [comment]
`BOOST_NO_NOEXCEPT`, documented in Boost.Config version 1.50:
http://www.boost.org/doc/libs/1_50_0/libs/config/doc/html/boost_config/boost_macro_reference.html

**Comment by GilesBathgate** [comment]
@lrineau ok, so it seems neither of these macros determines whether CXX11 noexcept is available, or have I missed something?

**Comment by GilesBathgate** [comment]
@lrineau My mistake, I was looking at BOOST_NOEXCEPT, not BOOST_NO_NOEXCEPT.

**Comment by lrineau** [comment]
Merge in [CGAL-Ic-73][i].

[i]: https://cgal.geometryfactory.com/CGAL/Members/testsuite/results-4.8-Ic-73.shtml

**Comment by GilesBathgate** [comment]
@lrineau Good news. Would it be possible to open up the test results you linked to non members?

**Comment by lrineau** [comment]
Unfortunately, for the moment that web site is password-protected. We plan to migrate to a Docker container for it, and then we will open it to public.

**Comment by mglisse** [comment]
Why not `CGAL_NOEXCEPT(CGAL_NO_ASSERTIONS)`, to avoid penalizing release builds?

**Comment by GilesBathgate** [comment]
@mglisse Sounds sensible, I assume you mean to amend calls to the macro, rather than the macro itself?

@lrineau Shall I push this change also, or does this interfere with the current test cycle?

**Comment by lrineau** [comment]
@GilesBathgate You can push a modification. It will be retested.

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-91

**Comment by sloriot** [comment]
`CGAL_NO_ASSERTIONS` is a macro keyword that is defined or not. That is you cannot use it as a boolean. You can add in `Installation/include/CGAL/config.h`:
````
    #ifndef CGAL_NO_ASSERTIONS
        #define CGAL_NO_ASSERTIONS_BOOL 0
    #else
       #define CGAL_NO_ASSERTIONS_BOOL 1
    #endif
````
and use it when using  `CGAL_NOEXCEPT`.

**Comment by lrineau** [comment]
That would be simpler to define `CGAL_NO_ASSERTIONS` to `1`, when it is defined.

**Comment by sloriot** [comment]
But then you're breaking the former usage of `CGAL_NO_ASSERTIONS`

**Comment by sloriot** [comment]
and this will not solve the problem if it is not defined...

**Comment by lrineau** [comment]
How would it be broken?! As long as the macro is never defined to 0, that makes no difference if `CGAL_NO_ASSERTIONS` is defined to 1 or to the empty string.

**Comment by GilesBathgate** [comment]
@lrineau Because I use CGAL_NO_ASSERTIONS in place of false in commit https://github.com/GilesBathgate/cgal/commit/0b11d8dd7dbf29a300d2ed00f272d0eb3676081a
e.g. CGAL_NOEXCEPT(CGAL_NO_ASSERTIONS) as suggested by @mglisse

**Comment by GilesBathgate** [comment]
@lrineau Furthermore I don't think an undefined macro evaluates to an empty string, it would give the compiler error: ‘CGAL_NO_ASSERTIONS’ was not declared in this scope

**Comment by lrineau** [comment]
Currently in CGAL-4.7, the macro is either undefined, or defined to an empty string:
```C++
#define CGAL_NO_ASSERTIONS
```

You can use an undefined macro in `#ifdef`/`#ifndef`, but also in `#if`: in conditional preprocessing directives, any unknown identifier is replaced by `0` before the expression is evaluated. But that is the only place I know where you can use an undefined macro (undefined identifier). And that is not true for a macro that is defined to the empty string. With the following code:
```C++
#define A
#if A
#endif
```
the compiler returns an error:
```
test.cpp:2:6: error: #if with no expression
 #if A
      ^
```

That is why I would like that, in CGAL, we define macro that way:
 - either the macro is left undefined,
 - or it is defined to `1`.

But, indeed, that would not be sufficient for you use `CGAL_NOEXCEPT(CGAL_NO_ASSERTIONS)`.

**Comment by GilesBathgate** [comment]
@lrineau Thanks that makes it clear. I think @sloriot solution is the cleanest in this situation.

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-99

**Pull request merged**